storage/range.go

```diff
@@ -452,32 +482,43 @@ func (r *Range) addReadOnlyCmd(method string, args proto.Request, reply proto.Re
 
 	// If read-consistency is set to INCONSISTENT, run directly.
 	if header.ReadConsistency == proto.INCONSISTENT {
+		// But disallow any inconsistent reads within txns.
+		if args.Header().ReadConsistency == proto.INCONSISTENT && args.Header().Txn != nil {
+			reply.Header().SetGoError(util.Errorf("cannot allow inconsistent reads within a transaction"))
+			return reply.Header().GoError()
+		}
 		return r.executeCmd(0, method, args, reply)
+	} else if args.Header().ReadConsistency == proto.CONSENSUS {
+		reply.Header().SetGoError(util.Errorf("consensus reads not implemented"))
+		return reply.Header().GoError()
+	}
+
+	held, prevTerm := r.HasLeaderLease()
+	if !held {
+		// If we don't hold the lease, but nobody else does either, submit
+		// this read-only command as read-write. The read will only succeed
+		// if this replica is the leader, and the command queue and read
+		// timestamp cache are both updated the same as if we had the lease.
+		if anotherHeld, _ := r.AnotherHasLeaderLease(); !anotherHeld {
```

Comment:
Lot of comments here. Let me try to address then all.

- We don't guarantee the other holds the lease. We return whether another replica has it or has held it and we also return a second bool which indicates whether it's expired.
- We use the clock offset in favor of the other holding it purposefully. The idea here is to give the other replica every consideration in time for it to complete a regrant before this replica decides to try to snag the lease itself. Whether or not the lease held by someone else is expired is not currently used, but will be when we respond to write/read pressure.
- I'd rather write exactly the same information to every replica then to have replicas start recording different expiration times for their leases depending on their local clock at the time the lease was received. Afterall, the time at which a lease might be received could be aeons after it's already expired in absolute time. Letting replicas diverge should be literally the worst thing we allow to happen--the sort of problems that would arise from making the wrong mistake relaxing that guarantee would be disastrous.
- There is no correctness issue with our use of clock offset with these leases. The leases are used in multiraft in such a way that guarantees a quorum can't be reached on a new election before the lease expires--this is not dependent on clock offset. Further, nodes believe correctly whether they hold the lease because the lease holder generates the expiration using its system clock.
- The issue which seems to be motivating your comment is clock offset between nodes as regards comparison with another lease holder's expiration. This is not a correctness issue. At some point in the future, we may decide to use this information to have a local replica start campaigning if there's significant read/write pressure. However, since it can't win an election until lease is definitively up, this isn't harmful, just possibly annoying. But, we don't even use the comparison of local system clock with another lease holder's expiration, so this is a moot point.
- Let's worry about silly lease durations when that becomes a problem. Right now it's hard coded as 1s and until that's configurable, no sense in putting in a check. When we find raft latencies significantly in excess of 1s, I agree it's probably time to rethink this, but planning for every contingency seems unnecessary.